### DISCUSSION TITLE
Raw advertisement data

### DISCUSSION THREAD

**Discussion by FiachraBarry99** [hbldh/bleak#540] [Q&A]
Is there a way to access the raw advertisement data that BleakScanner picks up?

**Comment by dlech** [comment]
The underlying OSes don't really do this. What in particular are you trying to do?

You can get the OS-specific data structure using the `platform_data` property of the advertising data object in the advertising data callback.

**Reply by FiachraBarry99** [reply]
Thanks for the quick reply! I was afraid you would say something like that. I am using Windows OS which seems, from my research, to be particularly unsuited to BLE applications. I am trying to get temperature data from a BLE sensor. It broadcasts the data in a standard format which can be used to read the temperature, humidity etc. The details of the sensor and its advertising data structure can be found [here](https://usermanual.wiki/Document/TemperatureHumidityDataLoggerCommandsAPI24.2837071165/html).

I'm a major noob when it comes to coding and even more so when it comes to BLE. How would I structure code to access the platform_data?

**Reply by FiachraBarry99** [reply]
I have the BLEDevice from the BleakScanner already.

**Reply by dlech** [reply]
It looks like it should show up as the manufacturer data. If it doesn't, then they probably didn't follow the Bluetooth specifications for formatting the advertising data correctly and there probably isn't a way to access it.

What is printed if you run the `detection_callback.py` example program while the device is advertising?

**Reply by FiachraBarry99** [reply]
```
C0:3D:AA:ED:C6:39 RSSI: -73 AdvertisementData(local_name='C03DAAED', manufacturer_data={307: b'\x17d\x0e\x10\x0e\xc4\x00\xf2\x01t\x00W\x01\x00'})
C0:3D:AA:ED:C6:39 RSSI: -73 AdvertisementData(manufacturer_data={307: b'\x02\xad\x02\x8d\x00\x9b\x00\x0e\x01\xc6\x01^\x00\xf3\x00h\x01<\x00\xe4\x00\x00\x00\x00\x00
```

That is the section with the MAC address I'm looking for.

**Reply by dlech** [reply]
Nice. So the `manufacturer_data` is what you are looking for. Their manufacturer ID is `307` (0x0133), so to access the raw data, use `adv.manufacturer_data[307]`. You can use `struct.unpack(...)` to extract the values. You can probably use the length of the data to figure out if it is the first or second part (the long one is bytes 3 to 30 and the short one is bytes 33 to 57).

**Reply by dlech** [reply]
Actually, the short one looks like the first part since the first byte is `\x17` and that matches the Version No. in the datasheet you linked.

**Reply by FiachraBarry99** [reply]
Sorry for all the questions. What do you mean use `adv.manufacturer_data[307]`?

**Reply by FiachraBarry99** [reply]
From my understanding of the docs the manufacturer data should be stored in BLEDevice.metadata. However, when I try that I get this: `{'uuids': [], 'manufacturer_data': {307: b''}}` .

**Reply by dlech** [reply]
Not sure if that is a bug in bleak or something to do with the device. Either way, `metadata` is "old" and shouldn't be used. Use `advertisement_data` from the detection callback as in the example `detection_callback.py`.

Then do something like this:

```python
import struct

BLUE_MAESTRO = 307  # company ID
VERSION = 0x17

def simple_callback(device: BLEDevice, advertisement_data: AdvertisementData):
    data = advertisement_data.manufacturer_data.get(BLUE_MAESTRO)
    if not data:
        return

    if data[0] == VERSION:  # would be safer to check len(data) == ? instead but I'm too lazy to count the bytes
       # advertisement packet
        version, battery, logging, ... = struct.unpack("<BBH...", data)
        # do stuff with values
    else:
        # scan response packet
        high_temp, high_humidity, ... = struct.unpack("<hh...", data)
        # do stuff with values
```

Replacing the `...`s is an exercise left to the reader.

**Reply by dlech** [reply]
Unless you actually need to connect to the device later, you can ignore the `BLEDevice` object.

**Reply by FiachraBarry99** [reply]
Would this essentially run indefinitely? I would prefer to only run once, scan until the advertising data from the device was read or a time limit was reached, store the manufacturer data, do the `struct.unpack` to locate temperature and convert to decimal and then return the temperature.

**Reply by dlech** [reply]
Sure, you can change the program however you like. `asyncio.Queue` is a nice way to pass data from the callback to somewhere else.

**Reply by FiachraBarry99** [reply]
Thanks so much for all the help. Think I should go away and get a proper handle on asyncio because I think I am just confusing myself.

Just to clarify `register_detection_callback` calls the callback function whenever the BleakScanner detects a new device or change in properties right?

**Reply by dlech** [reply]
Yes that is the idea.

**Reply by FiachraBarry99** [reply]
Hi, just want to say I got my script running. It scans for the particular device I'm looking for and returns the info from the device. Thank you so much for the help without it I would have been lost.

**Reply by Gvozdyara** [reply]
> Hi, just want to say I got my script running. It scans for the particular device I'm looking for and returns the info from the device. Thank you so much for the help without it I would have been lost.

Hi, could share your solution?

**Reply by FiachraBarry99** [reply]
@Gvozdyara you can check out my solution [here](https://github.com/FiachraBarry99/BlueMaestroAPI/blob/main/BlueMaestroAPI.py)
